File: src/com/github/unchama/seichiassist/data/Mana.java
package com.github.unchama.seichiassist.data;



import java.util.UUID;

import net.md_5.bungee.api.ChatColor;

import org.bukkit.boss.BarColor;
import org.bukkit.boss.BarStyle;
import org.bukkit.boss.BossBar;
import org.bukkit.entity.Player;

import com.github.unchama.seichiassist.SeichiAssist;
import com.github.unchama.seichiassist.util.Util;



public class Mana {
	//マナの値
	double m;
	//マックスの値
	double max;
	//バークラス
	BossBar manabar;
	//読み込まれているかどうかのフラグ
	boolean loadflag;

	//引数なしのコンストラクタ
	public Mana() {
		this.m = 0;
		this.max = 0;
		this.loadflag = false;
	}
	//必ず最初のみ実行してほしいメソッド
	public void update(Player player, int level) {
		//現在のレベルでのマナ上限値を計算しバーに表示
		//mの値は既に得られているはず。
		this.loadflag = true;
		this.calcMaxMana(player, level);
		displayMana(player,level);

	}
	//現在マナをマナバーに表示させる
	public void displayMana(Player player,int level){
		if(!loadflag)return;
		removeBar();
		setManaBar(player,level);
	}
	private void setManaBar(Player player,int level) {
		manabar = player.getServer().createBossBar(ChatColor.AQUA + "" + ChatColor.BOLD + "マナ(" + Util.Decimal(m) + "/" + max + ")", BarColor.BLUE, BarStyle.SOLID);

		if(m/max < 0.0 || m/max > 1.0){
			reset(player, level);
			player.sendMessage(ChatColor.RED + "不正な値がマナとして保存されていたためリセットしました。");
		}
		if (!(max == 0||max < 0)){
		manabar.setProgress(m/max);
		manabar.addPlayer(player);
		}
	}
	private void reset(Player player, int level) {
		this.calcMaxMana(player, level);
		if(this.m < 0.0)this.m = 0;
		if(this.m > max)this.m = max;
	}
	//現在のバーを削除する（更新するときは不要）
	public void removeBar(){
		try{manabar.removeAll();}catch(NullPointerException e){}
	}
	public void increaseMana(double i,Player player,int level){
		this.m += i;
		if(m > max) m = max;
		displayMana(player, level);
	}
	public void decreaseMana(double d,Player player,int level){
		this.m -= d;
		if(m < 0) m = 0;
		if(SeichiAssist.DEBUG)m = max;
		displayMana(player,level);
	}
	public boolean hasMana(double h){
        return compare(m, h) >= 0;
    }

	private int compare(double x, double y) {
		if(x > y)return 1;
		if(x == y)return 0;
		return -1;
	}
	//レベルアップするときに実行したい関数
	public void LevelUp(Player player,int level){
		calcMaxMana(player, level);
		fullMana(player,level);
	}
	//マナ最大値を計算する処理
	public void calcMaxMana(Player player, int level){

		//UUIDを取得
		UUID uuid = player.getUniqueId();
		//playerdataを取得
		PlayerData playerdata = SeichiAssist.playermap.get(uuid);

		if(SeichiAssist.DEBUG){
			max = 100000;
			return;
		}
		double t_max = 1;
		//レベルが10行っていない時レベルの値で処理を終了(修正:マナは0)
		if(level < 10){
			//this.max = level;
			this.max = 0.0;
			return;
		}
		//１０行ってる時の処理
		t_max = 100;
		int increase = 10;
		int inc_inc = 2;
		//１１以降の処理
		for(int i = 10 ; i < level;i++){
			if(i%10 == 0 && i<=110 && i != 10){
				increase += inc_inc;
				inc_inc *= 2;
			}
			t_max += increase;
		}
		//貢献度ptの上昇値
		t_max += playerdata.added_mana * SeichiAssist.config.getContributeAddedMana();

		this.max = t_max;
		return;
	}

	/**
	 * @param level
	 * レベル
	 *
	 * @return
	 * 最大マナ
	 */
	//マナ最大値を計算する処理
	public double calcMaxManaOnly(Player player, int level){

		//UUIDを取得
		UUID uuid = player.getUniqueId();
		//playerdataを取得
		PlayerData playerdata = SeichiAssist.playermap.get(uuid);

		/*
		if(SeichiAssist.DEBUG){
			max = 100000;
			return;
		}
		*/
		double t_max = 1;
		//レベルが10行っていない時レベルの値で処理を終了
		if(level < 10){
			//temp_max = level;
			max=0.0;
			return 0.0;
		}
		//１０行ってる時の処理
		t_max = 100;
		int increase = 10;
		int inc_inc = 2;
		//１１以降の処理
		for(int i = 10 ; i < level;i++){
			if(i%10 == 0 && i<=110 && i != 10){
				increase += inc_inc;
				inc_inc *= 2;
			}
			t_max += increase;
		}
		//貢献度ptの上昇値
		t_max += playerdata.added_mana * SeichiAssist.config.getContributeAddedMana();

		max=t_max;
		return t_max;
	}



	//マナを最大値まで回復する処理
	public void fullMana(Player player,int level){
		this.m = this.max;
		displayMana(player,level);
	}
	public void setMana(double m) {
		this.m = m;
	}
	public double getMana() {
		return this.m;
	}
	public boolean isloaded() {
		return this.loadflag;
	}
}

File: src/main/scala/com/github/unchama/seichiassist/data/Mana.scala
package com.github.unchama.seichiassist.data

import com.github.unchama.seichiassist.SeichiAssist
import com.github.unchama.seichiassist.util.TypeConverter
import net.md_5.bungee.api.ChatColor
import org.bukkit.boss.BarColor
import org.bukkit.boss.BarStyle
import org.bukkit.boss.BossBar
import org.bukkit.entity.Player

class Mana {
  //マナの値
  private var value = 0.0
  //マックスの値
  private var max = 0.0
  //バークラス
  private var bar: BossBar = _
  //読み込まれているかどうかのフラグ
  private var loadflag = false

  //必ず最初のみ実行してほしいメソッド
  def initialize(player: Player, level: Int): Unit = {
    //現在のレベルでのマナ上限値を計算しバーに表示
    //mの値は既に得られているはず。
    loadflag = true
    calcAndSetMax(player, level)
    display(player, level)
  }

  //現在マナをマナバーに表示させる
  def display(player: Player, level: Int): Unit = {
    if (!loadflag) return
    hide()
    setBar(player, level)
  }

  private def setBar(player: Player, level: Int): Unit = {
    bar = player.getServer.createBossBar(ChatColor.AQUA + "" + ChatColor.BOLD + "マナ(" + TypeConverter.Decimal(value) + "/" + getMax + ")", BarColor.BLUE, BarStyle.SOLID)
    val beforeResetMax = getMax
    if (value < 0 || beforeResetMax < 0 || value > beforeResetMax) {
      reset(player, level)
      player.sendMessage(ChatColor.RED + "不正な値がマナとして保存されていたためリセットしました。")
    }
    val trueMax = getMax
    if (trueMax <= 0) return
    bar.setProgress(value / trueMax)
    bar.addPlayer(player)
  }

  private def reset(player: Player, level: Int): Unit = {
    calcAndSetMax(player, level)
    if (value < 0.0) value = 0
    if (value > getMax) value = getMax
  }

  //現在のバーを削除する（更新するときは不要）
  def hide(): Unit = if (bar != null) bar.removeAll()

  private def ensureNotOverflow(): Unit = if (value > getMax) value = getMax

  def increase(amount: Double, player: Player, level: Int): Unit = {
    value += amount
    ensureNotOverflow()
    display(player, level)
  }

  def decrease(amount: Double, player: Player, level: Int): Unit = {
    value -= amount
    if (value < 0) value = 0
    if (SeichiAssist.DEBUG) value = getMax
    display(player, level)
  }

  def has(amount: Double): Boolean = value >= amount

  //レベルアップするときに実行したい関数
  def onLevelUp(player: Player, level: Int): Unit = {
    calcAndSetMax(player, level)
    setFull(player, level)
  }

  //マナ最大値を計算する処理
  def calcAndSetMax(player: Player, level: Int): Unit = {
    //UUIDを取得
    val uuid = player.getUniqueId
    //playerdataを取得
    val playerdata = SeichiAssist.playermap.getOrElse(uuid, null)
    if (SeichiAssist.DEBUG) {
      setMax(100000)
      return
    }
    //レベルが10行っていない時レベルの値で処理を終了(修正:マナは0)
    if (level < 10) { //this.max = level;
      setMax(0.0)
      return
    }
    //１０行ってる時の処理
    var t_max = 100
    var increase = 10
    var inc_inc = 2
    //１１以降の処理
    for (i <- 10 until level) {
      if (i % 10 == 0 && i <= 110 && i != 10) {
        increase += inc_inc
        inc_inc *= 2
      }
      t_max += increase
    }
    //貢献度ptの上昇値
    t_max += playerdata.added_mana * SeichiAssist.seichiAssistConfig.getContributeAddedMana
    setMax(t_max)
  }

  /**
   * @param level レベル
   * @return 最大マナ
   */
  //マナ最大値を計算する処理
  def calcMaxManaOnly(player: Player, level: Int): Double = {
    //UUIDを取得
    val uuid = player.getUniqueId
    //playerdataを取得
    val playerdata = SeichiAssist.playermap.getOrElse(uuid, null)
    /*
        if(SeichiAssist.DEBUG){
          max = 100000;
          return;
        }
        */
    //レベルが10行っていない時レベルの値で処理を終了
    if (level < 10) { //temp_max = level;
      setMax(0.0)
      return 0.0
    }
    //１０行ってる時の処理
    var t_max = 100
    var increase = 10
    var inc_inc = 2
    //１１以降の処理
    for (i <- 10 until level) {
      if (i % 10 == 0 && i <= 110 && i != 10) {
        increase += inc_inc
        inc_inc *= 2
      }
      t_max += increase
    }
    //貢献度ptの上昇値
    t_max += playerdata.added_mana * SeichiAssist.seichiAssistConfig.getContributeAddedMana
    setMax(t_max)
    t_max
  }

  //マナを最大値まで回復する処理
  def setFull(player: Player, level: Int): Unit = {
    value = getMax
    display(player, level)
  }

  def getMana: Double = value

  def setMana(m: Double): Unit = value = m

  def isLoaded: Boolean = loadflag

  def getMax: Double = max

  private def setMax(max: Double): Unit = this.max = max
}
